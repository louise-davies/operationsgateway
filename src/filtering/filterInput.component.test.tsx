import React from 'react';
import FilterInput from './filterInput.component';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('Filter input component', () => {
  let props: React.ComponentProps<typeof FilterInput>;

  beforeEach(() => {
    props = {
      error: '',
      value: [],
      setValue: jest.fn(),
      setError: jest.fn(),
      channels: ['timestamp', 'shotnum'],
    };
  });

  it('renders a autocomplete with chips for any values the user has selected', () => {
    props.value = [
      { type: 'channel', value: 'timestamp' },
      { type: 'unaryop', value: 'is not null' },
    ];
    const view = render(<FilterInput {...props} />);

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders an error if there is one', () => {
    props.error = 'Test error';
    const view = render(<FilterInput {...props} />);

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('user can type in channels to the filter', async () => {
    const user = userEvent.setup();
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, 'timestamp');
    await user.type(filter, '{enter}');

    expect(props.setValue).toHaveBeenCalledWith([
      { type: 'channel', value: 'timestamp' },
    ]);
    expect(props.setError).toHaveBeenCalledWith('');
  });

  it('user can type numbers to the filter', async () => {
    const user = userEvent.setup();
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, '1');
    await user.type(filter, '{enter}');

    expect(props.setValue).toHaveBeenCalledWith([
      { type: 'number', value: '1' },
    ]);
    expect(props.setError).toHaveBeenCalledWith('');
  });

  it('user can type custom strings to the filter if they are wrapped in double quotes', async () => {
    const user = userEvent.setup();
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, '"test"');
    await user.type(filter, '{enter}');

    expect(props.setValue).toHaveBeenCalledWith([
      { type: 'string', value: '"test"' },
    ]);
    expect(props.setError).toHaveBeenCalledWith('');
  });

  it('user can type custom strings to the filter if they are wrapped in single quotes', async () => {
    const user = userEvent.setup();
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, "'test'");
    await user.type(filter, '{enter}');

    expect(props.setValue).toHaveBeenCalledWith([
      { type: 'string', value: "'test'" },
    ]);
    expect(props.setError).toHaveBeenCalledWith('');
  });

  it("user can't type custom strings to the filter if they aren't wrapped in quotes", async () => {
    const user = userEvent.setup();
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, 'test');
    await user.type(filter, '{enter}');

    expect(props.setValue).not.toHaveBeenCalled();
  });

  it('validates the value when the user stops focusing on the input and sets an error if invalid', async () => {
    const user = userEvent.setup();
    props.value = [
      { type: 'channel', value: 'timestamp' },
      { type: 'unaryop', value: 'is not null' },
      { type: 'and', value: 'and' },
    ];
    render(<FilterInput {...props} />);

    const filter = screen.getByLabelText('Filter');

    await user.type(filter, 'timestamp');
    await user.type(filter, '{enter}');
    (props.setError as jest.Mock).mockClear();
    await user.tab();

    expect(props.setError).toHaveBeenCalledWith(
      // aka expect non-empty string
      expect.not.stringMatching(/^$/)
    );
  });

  it('validates the value when the user stops focusing on the input and clears an error if valid', async () => {
    const user = userEvent.setup();
    props.value = [
      { type: 'channel', value: 'timestamp' },
      { type: 'unaryop', value: 'is not null' },
    ];
    render(<FilterInput {...props} />);

    await user.tab();
    await user.tab();

    expect(props.setError).toHaveBeenCalledWith('');
  });
});
