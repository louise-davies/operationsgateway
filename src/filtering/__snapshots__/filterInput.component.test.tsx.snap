// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter input component > renders a autocomplete with chips for any values the user has selected 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon css-1u6ug4u-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Filter
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
        data-id="Input"
      >
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeSmall css-1a9jshs-MuiButtonBase-root-MuiChip-root"
          data-tag-index="0"
          role="button"
          tabindex="-1"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            type
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconSmall MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12yjm75-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeSmall css-1a9jshs-MuiButtonBase-root-MuiChip-root"
          data-tag-index="1"
          role="button"
          tabindex="-1"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            is not null
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconSmall MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Filter
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Filter input component > renders an error if there is one 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-fullWidth css-1u6ug4u-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r2:"
        id=":r2:-label"
      >
        Filter
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
        data-id="Input"
      >
        <input
          aria-autocomplete="list"
          aria-describedby=":r2:-helper-text"
          aria-expanded="false"
          aria-invalid="true"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12yjm75-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Filter
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
        id=":r2:-helper-text"
      >
        Test error
      </p>
    </div>
  </div>
</DocumentFragment>
`;
